kamu-adapter-graphql/src/queries/datasets.rs

```diff
@@ -103,6 +105,206 @@ impl Datasets {
+    ) -> Result<CreateDatasetFromSnapshotResult> {
+        use odf::serde::DatasetSnapshotDeserializer;
+
+        let snapshot = match dataset_snapshot_format {
+            MetadataManifestFormat::Yaml => {
+                let de = odf::serde::yaml::YamlDatasetSnapshotDeserializer;
+                match de.read_manifest(dataset_snapshot.as_bytes()) {
+                    Ok(snapshot) => snapshot,
+                    Err(e @ odf::serde::Error::SerdeError { .. }) => {
+                        return Ok(CreateDatasetFromSnapshotResult::Malformed(
+                            MetadataManifestMalformed {
+                                message: e.to_string(),
+                            },
+                        ))
+                    }
+                    Err(odf::serde::Error::UnsupportedVersion(e)) => {
+                        return Ok(CreateDatasetFromSnapshotResult::UnsupportedVersion(
+                            e.into(),
+                        ))
+                    }
+                    Err(e @ odf::serde::Error::IoError { .. }) => return Err(e.into()),
+                }
+            }
+        };
+
+        let local_repo = from_catalog::<dyn domain::LocalDatasetRepository>(ctx).unwrap();
+
+        let result = match local_repo.create_dataset_from_snapshot(snapshot).await {
+            Ok(result) => {
+                let dataset = Dataset::from_ref(ctx, &result.dataset_handle.as_local_ref()).await?;
+                CreateDatasetFromSnapshotResult::Success(CreateDatasetResultSuccess { dataset })
+            }
+            Err(domain::CreateDatasetFromSnapshotError::NameCollision(e)) => {
+                CreateDatasetFromSnapshotResult::NameCollision(CreateDatasetResultNameCollision {
+                    dataset_name: e.name.into(),
+                })
+            }
+            Err(domain::CreateDatasetFromSnapshotError::InvalidSnapshot(e)) => {
+                CreateDatasetFromSnapshotResult::InvalidSnapshot(
+                    CreateDatasetResultInvalidSnapshot { message: e.reason },
+                )
+            }
+            Err(domain::CreateDatasetFromSnapshotError::MissingInputs(e)) => {
+                CreateDatasetFromSnapshotResult::MissingInputs(CreateDatasetResultMissingInputs {
+                    missing_inputs: e
+                        .missing_inputs
+                        .into_iter()
+                        .map(|r| r.to_string())
+                        .collect(),
+                })
+            }
+            Err(domain::CreateDatasetFromSnapshotError::Internal(e)) => return Err(e.into()),
+        };
+
+        Ok(result)
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// CreateDatasetResult
+///////////////////////////////////////////////////////////////////////////////
+
+#[derive(Interface, Debug, Clone)]
+#[graphql(field(name = "message", type = "String"))]
+pub enum CreateDatasetResult {
```

Comment:
Note the new approach to error handling that I'm suggesting:
* Result is an interface with single field `message` - this allows for a "catch-all" error handling code
* Success and individual errors can be unpacked as `... on CreateDatasetResultSuccess { dataset { id } }` so that detailed fields can be accessed
* This can be extended with an error code if we need it, but `_typename` should already serve as a unique error identifier

What do you think?

Reply:
UI only needs to display an error. Most of the time it would not need any complex polymorphic behavior.